nutils/pointsseq.py

```diff
@@ -0,0 +1,648 @@
+
+    Returns
+    -------
+    sequence : :class:`PointsSequence`
+        This product sequence.
+    '''
+
+    return Product(self, other)
+
+  def chain(self, other):
+    '''Return the chained sequence of this sequence with ``other``.
+
+    Parameters
+    ----------
+    other : :class:`PointsSequence`
+
+    Returns
+    -------
+    sequence : :class:`PointsSequence`
+        The chained sequence.
+    '''
+
+    if not other:
+      return self
+    elif not self:
+      return other
+    elif self == other:
+      return Repeat(self, 2)
+    else:
+      return Chain(self, other).balanced
+
+  @property
+  def isuniform(self):
+    '''``True`` if all points in this sequence are equal.'''
+
+    return len(self) == 1
+
+  @property
+  def tri(self):
+    '''Triangulation of interior.
+
+    A two-dimensional integer array with ``ndims+1`` columns, of which every
+    row defines a simplex by mapping vertices into the list of points.
+    '''
+
+    tri = []
+    offset = 0
+    for points in self:
+      tri.append(points.tri + offset)
+      offset += points.npoints
+    return types.frozenarray(numpy.concatenate(tri) if tri else numpy.zeros((0,self.ndims+1), int), copy=False)
+
+  @property
+  def hull(self):
+    '''Triangulation of the exterior hull.
+
+    A two-dimensional integer array with ``ndims`` columns, of which every row
+    defines a simplex by mapping vertices into the list of points. Note that
+    the hull often does contain internal element boundaries as the
+    triangulations originating from separate elements are disconnected.
+    '''
+
+    hull = []
+    offset = 0
+    for points in self:
+      hull.append(points.hull + offset)
+      offset += points.npoints
+    return types.frozenarray(numpy.concatenate(hull) if hull else numpy.zeros((0,self.ndims), int), copy=False)
+
+strictpointsseq = types.strict[PointsSequence] # type: ignore
+
+class Empty(PointsSequence):
+  '''An empty sequence.'''
+
+  __slots__ = ()
+
+  def __len__(self):
+    return 0
+
+  def get(self, index):
+    raise IndexError('sequence index out of range')
+
+class Plain(PointsSequence):
+  '''A general purpose implementation of :class:`PointsSequence`.
+
+  Use this class only if there exists no specific implementation of
+  :class:`PointsSequence` for the points at hand.
+
+  Parameters
+  ----------
+  points : :class:`tuple` of :class:`~nutils.points.Points` objects
+      The sequence of points.
+  ndims : :class:`int`
+      The dimension of the manifold space.
+  '''
+
+  __slots__ = '_items'
+
+  @types.apply_annotations
+  def __init__(self, items: types.tuple[points.strictpoints], ndims: types.strictint):
+    if len(items) == 0:
+      warnings.warn('expected at least 2 items but got 0; use `Empty(ndims)` instead', warnings.NutilsInefficiencyWarning)
+    elif len(items) == 1:
+      warnings.warn('expected at least 2 items but got 1; use `Uniform(items[0], 1)` instead', warnings.NutilsInefficiencyWarning)
```

Comment:
should we check instead if all elements are equal, since this situation should have been steered to Uniform by aspoints? (same holds in elementsseq)

Reply:
done